fuel-relayer/src/relayer.rs

```diff
@@ -0,0 +1,726 @@
+use std::{
+    cmp::{max, min},
+    time::Duration,
+};
+
+use crate::{
+    log::EthEventLog,
+    pending_events::{PendingDiff, PendingEvents},
+    validator_set::CurrentValidatorSet,
+    Config,
+};
+use futures::Future;
+use tokio::sync::Mutex;
+use tokio::sync::{broadcast, mpsc};
+use tracing::{error, info, trace, warn};
+
+use anyhow::Error;
+use ethers_core::types::{BlockId, Filter, Log, TxHash, ValueOrArray, H256};
+use ethers_providers::{
+    FilterWatcher, Middleware, Provider, ProviderError, StreamExt, SyncingStatus, Ws,
+};
+use fuel_core_interfaces::{
+    block_importer::NewBlockEvent,
+    relayer::{RelayerDb, RelayerError, RelayerEvent, RelayerStatus},
+    signer::Signer,
+};
+
+const REPORT_PROGRESS_EVERY_N_BLOCKS: u64 = 500;
+const PROVIDER_INTERVAL: u64 = 1000;
+///
+pub struct Relayer {
+    /// Pendning stakes/assets/withdrawals. Before they are finalized
+    pending: PendingEvents,
+    /// Current validator set
+    current_validator_set: CurrentValidatorSet,
+    /// db connector to apply stake and token deposit
+    db: Box<Mutex<dyn RelayerDb>>,
```

Comment:
I think its better to make the `Mutex` protection of the DB encapsulated there instead of forcing the consumers to do it themselves. Many dbs uses that approach. Ex: [sled](https://docs.rs/sled/0.34.7/src/sled/db.rs.html#9-14)

Reply:
It's also what we do inside of fuel-core, the Database is Send + Sync and handles any of the locking concerns internally if needed.

Reply:
Agree, i will make RelayerDb trait require `Send+Sync` as some functions in `Storage` are mut and because of that RelayerDb needs to have some mut functions.